### PR TITLE
206 handling table headers

### PR DESCRIPTION
**Pull request by OsamaMSoliman**
I need @vijulshah to check if the problem of #182 is still there or not before finalizing this PR!

### CODE REVIEW

**Comment by OsamaMSoliman** [comment]
start working on [#228](https://github.com/mlgym/mlgym/issues/228) here

File: src/ml_board/frontend/Dashboard/src/components/dashboard/table/Table.tsx [hunk 1]
```diff
@@ -41,12 +42,13 @@ export default function Table({ colNames, rows }: { colNames: string[], rows: an
         defaultColDef={defaultColDef}
         columnDefs={colDefs}
         rowData={rows}
-        onGridReady={params => params.api.sizeColumnsToFit()}
-        getRowId={(params: any) => params.data.job_id}
+        onGridReady={onGridReady}
+        getRowId={getRowId}
         rowSelection={"multiple"}
         onRowDoubleClicked={onRowDoubleClicked}
```
Comment (vijulshah) [inline src/ml_board/frontend/Dashboard/src/components/dashboard/table/Table.tsx:48]: Please remove double click

File: src/ml_board/frontend/Dashboard/src/components/dashboard/table/Table.tsx [hunk 2]
```diff
@@ -41,12 +42,13 @@ export default function Table({ colNames, rows }: { colNames: string[], rows: an
...
         rowStyle={{ cursor: "pointer" }}
         animateRows={true} // Optional - set to 'true' to have rows animate when sorted
+        enableCellChangeFlash={true}
```
Comment (vijulshah) [inline src/ml_board/frontend/Dashboard/src/components/dashboard/table/Table.tsx:51]: Flashing the cell on update... nice!!

File: src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx [hunk 3]
```diff
@@ -90,7 +90,6 @@ const Settings: React.FC<FuncProps> = (props) => {
 
     return (
         <Box>
-            <Toolbar/>
```
Comment (vijulshah) [inline src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx:93]: Pls, bring this back... as commented above

File: src/ml_board/frontend/Dashboard/src/worker_socket/DataTypes.ts [hunk 4]
```diff
@@ -20,7 +20,14 @@ export interface BufferedDataFromSocket extends JSON {
 }
 
 export interface DataToRedux {
+    tableHeaders?: string[],
     tableData: Row[],
     chartsUpdates: ChartUpdate[],
     status?: any,
+}
+
+export interface UpdatesObject {
```
Comment (vijulshah) [inline src/ml_board/frontend/Dashboard/src/worker_socket/DataTypes.ts:29]: can be renamed to something better... like DashboardTableObjects or maybe something else

File: src/ml_board/frontend/Dashboard/src/worker_socket/MainThreadCallbacks.ts
Comment (vijulshah) [inline src/ml_board/frontend/Dashboard/src/worker_socket/MainThreadCallbacks.ts:1]: Nice job regarding the logic of headers storage!! And also for the buffered msgs!!

**Review by vijulshah — CHANGES_REQUESTED** [review]
Everything looks good. Just 1 or 2 small updates as mentioned

File: src/ml_board/frontend/Dashboard/src/app/App.tsx [hunk 5]
```diff
@@ -155,22 +152,26 @@ export default function App() {
...
 
+    function dispatchTableHeadersUpdate(tableHeaders: Array<string> | undefined) { tableHeaders && dispatch(upsertTableHeaders(tableHeaders)); }
+
     return (
         <div className={styles.main_container}>
             {
                 // Show TopBar only if valid url is there. For example, if we get unregistered url (i.e 404 error) then don't show the TopBar
                 urls.includes(location.pathname.split("/")[1]) && <TopBarWithDrawer />
             }
+            <Toolbar/> {/* This acts as a padding buffer for the area under the TopBarWithDrawer and nothing more!  */}
```
[the remark this answers is not in the document]
Comment (OsamaMSoliman) [reply src/ml_board/frontend/Dashboard/src/app/App.tsx:174]: image

it's just an empty space that doesn't affect the 404 page! 🙈 I don't get it. what's the point? 😅😅

File: src/ml_board/frontend/Dashboard/src/components/dashboard/table/Table.tsx (hunk 1 above)
Comment (OsamaMSoliman) [reply src/ml_board/frontend/Dashboard/src/components/dashboard/table/Table.tsx:48]: done 👍

File: src/ml_board/frontend/Dashboard/src/worker_socket/DataTypes.ts (hunk 4 above)
Comment (OsamaMSoliman) [reply src/ml_board/frontend/Dashboard/src/worker_socket/DataTypes.ts:29]: but it doesn't just hold the updates of the Dashboard, it actually the graph updates too ?! 
https://github.com/mlgym/mlgym/blob/67b58a7ccf8b0d5423231f4094b5891799943e53/src/ml_board/frontend/Dashboard/src/worker_socket/DataTypes.ts#L29-L33

this is later processed into a `DataToRedux` object
https://github.com/mlgym/mlgym/blob/67b58a7ccf8b0d5423231f4094b5891799943e53/src/ml_board/frontend/Dashboard/src/worker_socket/DataTypes.ts#L22-L27

using this
https://github.com/mlgym/mlgym/blob/67b58a7ccf8b0d5423231f4094b5891799943e53/src/ml_board/frontend/Dashboard/src/worker_socket/MainThreadCallbacks.ts#L65-L74

**Comment by OsamaMSoliman** [comment]
After solving the merge conflicts, I find that the filter FAB is appearing in all pages, I'll debug later and ask for another review

File: src/ml_board/frontend/Dashboard/src/app/App.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (vijulshah) [reply src/ml_board/frontend/Dashboard/src/app/App.tsx:174]: You are right, but it's just good practice not to include things where they shouldn't be rendered

File: src/ml_board/frontend/Dashboard/src/worker_socket/DataTypes.ts (hunk 4 above)
Comment (vijulshah) [reply src/ml_board/frontend/Dashboard/src/worker_socket/DataTypes.ts:29]: Okay got it!

**Comment by vijulshah** [comment]
To be honest, the arrows on the left doesn't look good to me.... I think just the click on the row to navigate to the experiment page was quite good enough. (But whatever the team says, I am fine with it)

Issue-1:
I am filtering some headers -> then navigating inside an experiment -> coming back to the Dashboard page -> Now the headers which I had previously selected (filtered) are gone (reset) and again all the headers are being selected.
The same goes for text search also. 

Ideally, as a user - if I am using the functionality, I would need that the headers which I already filtered. should be kept as long as I am still using the app.
So I think a good solution should be to keep the selected headers (or their ids) in redux. Similarly, keep the search text in redux and the related selected headers in redux.

Issue-2:
Also, when all the headers are selected -> then if I filter and keep 5 or 6 headers -> the width of the columns doesn't expand automatically (i.e resizing columns/table itself is not auto-responsive)

(If this might be more work then let's make these as separate issues & maybe work on it later)

**Review by le1nux — APPROVED** [review]

**Pull request merged**